### PR TITLE
Add "join table" for visits

### PR DESCRIPTION
**Pull request by ONJoseph**
- [ ] Hi Reviewer, In this part the following were implemented:

- [ ] Create a table named vets with the following columns:
id: integer (set it as autoincremented PRIMARY KEY)
name: string
age: integer
date_of_graduation: date
There is a many-to-many relationship between the tables species and vets: a vet can specialize in multiple species, and a species can have multiple vets specialized in it. - [ ] Create a "join table" called specializations to handle this relationship.
There is a many-to-many relationship between the tables animals and vets: an animal can visit multiple vets and one vet can be visited by multiple animals. Create a "join table" called visits to handle this relationship, it should also keep track of the date of the visit.
- [ ] Insert the following data for vets:
Vet William Tatcher is 45 years old and graduated Apr 23rd, 2000.
Vet Maisy Smith is 26 years old and graduated Jan 17th, 2019.
Vet Stephanie Mendez is 64 years old and graduated May 4th, 1981.
Vet Jack Harkness is 38 years old and graduated Jun 8th, 2008.
Insert the following data for specialties:
Vet William Tatcher is specialized in Pokemon.
Vet Stephanie Mendez is specialized in Digimon and Pokemon.
Vet Jack Harkness is specialized in Digimon.
- [ ] Insert the following data for visits:
Agumon visited William Tatcher on May 24th, 2020.
Agumon visited Stephanie Mendez on Jul 22th, 2020.
Gabumon visited Jack Harkness on Feb 2nd, 2021.
Pikachu visited Maisy Smith on Jan 5th, 2020.
Pikachu visited Maisy Smith on Mar 8th, 2020.
Pikachu visited Maisy Smith on May 14th, 2020.
Devimon visited Stephanie Mendez on May 4th, 2021.
Charmander visited Jack Harkness on Feb 24th, 2021.
Plantmon visited Maisy Smith on Dec 21st, 2019.
Plantmon visited William Tatcher on Aug 10th, 2020.
Plantmon visited Maisy Smith on Apr 7th, 2021.
Squirtle visited Stephanie Mendez on Sep 29th, 2019.
Angemon visited Jack Harkness on Oct 3rd, 2020.
Angemon visited Jack Harkness on Nov 4th, 2020.
Boarmon visited Maisy Smith on Jan 24th, 2019.
Boarmon visited Maisy Smith on May 15th, 2019.
Boarmon visited Maisy Smith on Feb 27th, 2020.
Boarmon visited Maisy Smith on Aug 3rd, 2020.
Blossom visited Stephanie Mendez on May 24th, 2020.
Blossom visited William Tatcher on Jan 11th, 2021.
- [ ] Write queries to answer the following:
Who was the last animal seen by William Tatcher?
How many different animals did Stephanie Mendez see?
List all vets and their specialties, including vets with no specialties.
List all animals that visited Stephanie Mendez between April 1st and August 30th, 2020.
What animal has the most visits to vets?
Who was Maisy Smith's first visit?
Details for most recent visit: animal information, vet information, and date of visit.
How many visits were with a vet that did not specialize in that animal's species?
What specialty should Maisy Smith consider getting? Look for the species she gets the most.
- [ ] Take a screenshot of the results of your queries.
- [ ] Add the screenshot with the results of your queries to your Pull Request description.
Screenshot (56)
Screenshot (57)
Screenshot (58)
Screenshot (59)
Screenshot (60)
Screenshot (61)
Screenshot (62)
Screenshot (63)
Screenshot (64)](https://user-images.githubusercontent.com/60672480/222873696-ea56d240-3d40-4e1d-b31b-4b3a3e843ec1.png)
om/60672480/222873678-364734a4-f23d-4b44-a402-3344d642dd6a.png)

- [ ] Thanks!

### CODE REVIEW

File: schema.sql [hunk 1]
```diff
@@ -27,3 +27,31 @@ ALTER TABLE animals DROP COLUMN species;
...
+
+-- Create a "join table" called specializations
+CREATE TABLE specializations (
+spec_id INTEGER,
+vet_id INTEGER,
+CONSTRAINT fk_vets FOREIGN KEY (vet_id) REFERENCES vets(id),
+CONSTRAINT fk_species FOREIGN KEY (spec_id) REFERENCES species(id)
+);
+SELECT * FROM specializations
+
```
Comment (Owusu-Desmond) [inline schema.sql]: - [x] Please make sure that everything that is a query that helps to analyze data should be saved in the `queries.sql` file. :pray:

File: schema.sql [hunk 2]
```diff
@@ -27,3 +27,31 @@ ALTER TABLE animals DROP COLUMN species;
...
+-- Create a "join table" called visits
+CREATE TABLE visits (
+animal_id INTEGER,
+vet_id INTEGER,
+date_of_visit DATE,
+CONSTRAINT fk_animals FOREIGN KEY (animal_id) REFERENCES animals(id),
+CONSTRAINT fk_vets FOREIGN KEY (vet_id) REFERENCES vets(id)
+);
+SELECT * FROM visits;
```
Comment (Owusu-Desmond) [inline schema.sql:48]: - [x] Please make sure that everything that is a query that helps to analyze data should be saved in the `queries.sql` file. :pray:

File: data.sql [hunk 3]
```diff
@@ -47,3 +47,181 @@ INSERT INTO animals (name, species_id, owner_id) VALUES
...
+
+-- Insert the following data for specialties
+BEGIN;
+INSERT INTO specializations (vet_id, spec_id)
+SELECT vets.id, species.id FROM vets, species
+WHERE vets.name = 'William Tatcher'
+AND species.name = 'Pokemon';
+COMMIT;
+SELECT * FROM specializations;
+
```
Comment (Owusu-Desmond) [inline data.sql:66]: - [x] Please make sure that everything that is a query that helps to analyze data should be saved in the `queries.sql` file. :pray:

File: data.sql [hunk 4]
```diff
@@ -47,3 +47,181 @@ INSERT INTO animals (name, species_id, owner_id) VALUES
...
+BEGIN;
+INSERT INTO specializations (vet_id, spec_id)
+SELECT vets.id, species.id FROM vets, species
+WHERE vets.name = 'Stephanie Mendez'
+AND species.name IN ('Digimon','Pokemon');
+COMMIT;
+SELECT * FROM specializations;
+
```
Comment (Owusu-Desmond) [inline data.sql:74]: - [x] Please make sure that everything that is a query that helps to analyze data should be saved in the `queries.sql` file. :pray:

File: data.sql [hunk 5]
```diff
@@ -47,3 +47,181 @@ INSERT INTO animals (name, species_id, owner_id) VALUES
...
+BEGIN;
+INSERT INTO specializations (vet_id, spec_id)
+SELECT vets.id, species.id FROM vets, species
+WHERE vets.name = 'Jack Harkness'
+AND species.name = 'Digimon';
+COMMIT;
+SELECT * FROM specializations
+
```
Comment (Owusu-Desmond) [inline data.sql]: - [x] Please make sure that everything that is a query that helps to analyze data should be saved in the `queries.sql` file. :pray:

File: data.sql [hunk 6]
```diff
@@ -47,3 +47,181 @@ INSERT INTO animals (name, species_id, owner_id) VALUES
...
+SELECT animals.id, vets.id, '2020-05-24' FROM animals, vets
+WHERE animals.name = 'Blossom'
+AND vets.name = 'Stephanie Mendez';
+
+INSERT INTO visits (animal_id, vet_id, date_of_visit)
+SELECT animals.id, vets.id, '2021-01-11' FROM animals, vets
+WHERE animals.name = 'Blossom'
+AND vets.name = 'William Tatcher';
+COMMIT;
+SELECT * FROM visits;
```
Comment (Owusu-Desmond) [inline data.sql:186]: - [x] Please make sure that everything that is a query that helps to analyze data should be saved in the `queries.sql` file. :pray:

File: data.sql [hunk 7]
```diff
@@ -47,3 +47,181 @@ INSERT INTO animals (name, species_id, owner_id) VALUES
...
+SET owner_id = (SELECT id FROM owners WHERE full_name = 'Melody Pond')
+WHERE name IN ('Charmander', 'Squirtle', 'Blossom');
+
+UPDATE animals
+SET owner_id = (SELECT id FROM owners WHERE full_name = 'Dean Winchester')
+WHERE name IN ('Angemon', 'Boarmon');
+COMMIT;
+SELECT * from animals;
+ROLLBACK;
+
```
Comment (Owusu-Desmond) [inline data.sql]: - [x] Please make sure that everything that is a query that helps to analyze data should be saved in the `queries.sql` file. :pray:

**Review by Owusu-Desmond — CHANGES_REQUESTED** [review]
Hi @ONJoseph , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!



## Highlights
📌 No Linters error ✅
📌 Descriptive pull request✅
📌 Good commit messages ✅
📌 Professional README file ✅

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: data.sql (hunk 7 above)
Comment (ONJoseph) [reply data.sql]: @Owusu-Desmond Thanks a lot for your kind review! All requested changes implemented. Cheers!

File: queries.sql [hunk 8]
```diff
@@ -88,6 +88,252 @@ SELECT animals.name FROM animals JOIN owners ON animals.owner_id = owners.id WHE
...
+SELECT * FROM specializations;
+
+-- Insert the following data for visits:
+BEGIN;
+INSERT INTO visits (animal_id, vet_id, date_of_visit)
+SELECT animals.id, vets.id, '2020-05-24' FROM animals, vets
+WHERE animals.name = 'Agumon'
+AND vets.name = 'William Tatcher';
+COMMIT;
+
```
Comment (victorokoroji) [inline queries.sql]: - [x] Please ensure that everything that inserts data into the database is added to the data.sql file and not the queries.sql file. Kindly remove the data insertion lines 👍🏽 . 🙏.

File: queries.sql [hunk 9]
```diff
@@ -88,6 +88,252 @@ SELECT animals.name FROM animals JOIN owners ON animals.owner_id = owners.id WHE
...
+);
+SELECT * FROM visits;
+
+-- Create a "join table" called specializations
+CREATE TABLE specializations (
+spec_id INTEGER,
+vet_id INTEGER,
+CONSTRAINT fk_vets FOREIGN KEY (vet_id) REFERENCES vets(id),
+CONSTRAINT fk_species FOREIGN KEY (spec_id) REFERENCES species(id)
+);
```
Comment (victorokoroji) [inline queries.sql]: - [x] Please ensure that everything that creates a table is added to the schema.sql file and not the queries.sql file. Kindly remove the data insertion lines 👍🏽 . 🙏.

**Review by victorokoroji — CHANGES_REQUESTED** [review]
Hi  @ONJoseph, 

Great work on making the changes requested by a previous reviewer 👏 🔥 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.



## Top highlight 🎯

- The PR title and description are professional and descriptive ✅ 
- The commit messages are meaningful ✅ 


 ## Required Changes ♻️

- [x] Please ensure that everything that inserts data into the database is added to the data.sql file and not the queries.sql file. Kindly remove the data insertion lines 👍🏽 . 🙏.

- [x] Please ensure that everything that creates a table is added to the schema.sql file and not the queries.sql file. Kindly remove the data insertion lines 👍🏽 . 🙏.

- [x] As suggested by the previous reviewer, Please make sure that everything that is a query that helps to analyze data should be saved in the queries.sql file and not the data.sql file. 🙏

_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: queries.sql (hunk 9 above)
Comment (ONJoseph) [reply queries.sql]: Done. Thanks.

**Review by sumairq — APPROVED** [review]
### Hi @ONJoseph , 



Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
✔️ Good commit history
✔️Linters are passing
✔️Very professional README
✔️Correct Git flow
✔️Descriptive PR
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**